src/modm/ui/color/concepts.hpp

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (c) 2021, Thomas Sommer
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#pragma once
+
+#include <bit>
+
+/**
+ * @brief		Concepts for colortypes and groups of colortypes.
+ * 				Key applications:
+ * 				- Color conversion constructors
+ * 				- Specialisation of buffer manipulation algorithms
+ *
+ * @author		Thomas Sommer
+ * @ingroup		modm_ui_color
+ */
+namespace modm::color {
+
+template <int>
+class GrayD;
+
+template<int, int, int>
+class RgbD;
+
+template<int, int, int>
+class RgbStackedD;
+
+template <int, int, int>
+class HsvD;
+
+/**
+ * @brief 	Identify template class instance
+ * @see:	https://stackoverflow.com/questions/44012938/how-to-tell-if-template-type-is-an-instance-of-a-template-class
+ */
+template <class, template <int...> class>
+struct is_instance : public std::false_type {};
+
+template <int...Ts, template <int...> class U>
+struct is_instance<U<Ts...>, U> : public std::true_type {};
+
+/**
+ * @brief 	Concepts to ident specific colortype instance
+ */
+template<class C>
+concept ColorGray = is_instance<C, GrayD>::value;
+
+template<class C>
+concept ColorRgb = is_instance<C, RgbD>::value;
+
+template<class C>
+concept ColorRgbStacked = is_instance<C, RgbStackedD>::value;
+
+template<class C>
+concept ColorHsv = is_instance<C, HsvD>::value;
+
+/**
+ * @brief 	Concept to ident any colortype instance
+ */
+template<class C>
+concept Color = ColorGray<C> || ColorRgb<C> || ColorHsv<C> || ColorRgbStacked<C>; // conjunction
```

Comment:
An implementation of a `Color class` will be of specific nature, hence will have a name different from just "Color" like Rgb, Hsv a.s.o. "Color" [complements the inherent nature](https://en.cppreference.com/w/cpp/language/constraints) of _concepts_ and makes a quite precise name for the application!

> The intent of concepts is to model semantic categories (Number, Range, RegularFunction) rather than syntactic restrictions (HasPlus, Array). According to ISO C++ core guideline T.20, "The ability to specify meaningful semantics is a defining characteristic of a true concept, as opposed to a syntactic constraint."

**I hope** redefinition of plain "Color" as a class is permited ... aaaaaaand. ⏰ ... it is 😊:

> error: 'struct modm::color::Color' redeclared as different kind of entity